tools/sdk-generation-pipeline/packages/sdk-generation-lib/src/lib/ArtifactUploader.ts

```diff
@@ -0,0 +1,189 @@
+        this.validateContext(artifactBlobUploaderContext);
+        this.azureBlobClient = new AzureBlobClient(
+            artifactBlobUploaderContext.azureStorageBlobSasUrl,
+            artifactBlobUploaderContext.azureBlobContainerName
+        );
+        this.language = artifactBlobUploaderContext.language;
+        this.pipelineBuildId = artifactBlobUploaderContext.pipelineBuildId;
+    }
+
+    private validateContext(artifactBlobUploaderContext: ArtifactBlobUploaderContext) {
+        if (!artifactBlobUploaderContext.azureStorageBlobSasUrl) {
+            throw new Error(`Invalid azureStorageBlobSasUrl`);
+        }
+        if (!artifactBlobUploaderContext.azureBlobContainerName) {
+            throw new Error(`Invalid azureBlobContainerName`);
+        }
+        if (!(<any>Object).values(SDK).includes(artifactBlobUploaderContext.language)) {
+            throw new Error(`Invalid language`);
+        }
+        if (!artifactBlobUploaderContext.pipelineBuildId) {
+            throw new Error(`Invalid pipelineBuildId`);
+        }
+    }
+
+    public async uploadSourceCode(generateAndBuildOutputJson: GenerateAndBuildOutput) {
+        for (const p of generateAndBuildOutputJson.packages) {
+            const result = p.result;
+            if (result === 'failed') {
+                logger.warn(`Build ${p.packageName} failed, skipped it`);
+                continue;
+            }
+            const packageName = p.packageName;
+            const packageFolder = p.packageFolder;
+
+            if (packageFolder && fs.existsSync(packageFolder)) {
+                for (const filePath of getFileListInPackageFolder(packageFolder)) {
+                    if (fs.existsSync(path.join(packageFolder, filePath))) {
+                        await this.azureBlobClient.publishBlob(
+                            path.join(packageFolder, filePath),
+                            `${this.pipelineBuildId}/${this.language}/${packageName}/${filePath}`
+                        );
+                    }
+                }
+            }
+        }
+    }
+
+    public async uploadArtifacts(generateAndBuildOutputJson: GenerateAndBuildOutput) {
+        for (const p of generateAndBuildOutputJson.packages) {
+            const result = p.result;
+            if (result === 'failed') {
+                logger.warn(`Build ${p.packageName} failed, skipped it`);
+                continue;
+            }
+            const artifacts = p.artifacts;
+            if (!artifacts) {
+                // artifacts is optional
+                continue;
+            }
+
+            for (const artifact of artifacts) {
+                const artifactName = path.basename(artifact);
+                await this.azureBlobClient.publishBlob(
+                    artifact,
+                    `${this.pipelineBuildId}/${this.language}/${artifactName}`
+                );
+            }
+        }
+    }
+}
+
+export type ArtifactPipelineUploaderContext = {
```

Comment:
Why do we need to write code to upload artifacts to the pipeline? We should instead just put the files into a output directory and use the existing DevOps Publish pipeline artifacts (https://docs.microsoft.com/en-us/azure/devops/pipelines/publish-pipeline-artifact?view=azure-devops&tabs=yaml). They already have a very optimized tool that will upload artifacts.

Reply:
Thx for the suggestion. Yes, I totally agree with you to use the DevOps Publish pipeline artifacts. In fact I already use it in my  local repo. I will push the pipeline yaml files in the next pull request.

Reply:
Keep this principal in mind for other aspects of this tool. I suspect we might also be able to use azcopy in the pipeline or shell script instead of manually writing code to upload blobs. We should only write code for things that don't already exist in some other form.